[CORE-13424] ct/reconciler: store LRO in CTP STM

rockwotj opened the pull request:
Rework the cloud topics reconciler to introduce an abstraction for "source", which a source is L0.
The abstraction allows us to easily mock all the L0 infrastructure that depends on annoying to setup
machinary like cluster::partition, ct::data_plane_api, etc.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

rockwotj committed:
ct/reconciler: add an abstraction for the reconciliation_source

This will enable testing of the reconciler, which will come in the next
commit.

Additionally, this also adds checkpointing using the ctp stm to bookmark
the last reconciled offset, instead of the fake offset we used before.

rockwotj committed:
ct/reconciler: add unit test

Add unit tests for the reconciler. This also catches a bug in the
previous commit where we incorrectly updated LRO using
`next_offset`, which are two different things. This all works out
because LRO starts at `kafka::offset::min()`, so `kafka::next_offset(o)`
in the reader config will return `0` for an empty partition.

rockwotj committed:
ct/reconciler: abort processing early if aborted

This *could* be the cause of the shutdown hangs we've seen, if
reconcilation takes some time.

There are other places (like syncing STMs) were we could be more robust
about aborting stuff instead of having hardcoded timeouts, but I'll have
to fix that in another commit.

dotnwat commented on src/v/cloud_topics/reconciler/reconciliation_source.h:
```diff
@@ -0,0 +1,85 @@
+    virtual ss::future<std::expected<void, errc>>
+    set_last_reconciled_offset(kafka::offset, ss::abort_source&) = 0;
+
+    // Create a reader for the reconciliation source, data should only be read
+    // above `last_reconciled_offset`.
```
just wondering, if reconciler has no other need for LRO than to make the reader, we could remove `last_reconciled_offset` interface?

rockwotj replied:
Oh we could. We could make this method just take a max size and an abort source. Do you think that is a better interface?

dotnwat replied:
maybe from a purist perspective. but practically speaking i don't think it matters much.

dotnwat commented:
compilation errors

rockwotj commented:
Force push: fix compilation errors (sigh) I forgot to remove a parameter from the construct service in a late refactor

wdberkeley commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -458,28 +435,38 @@ reconciler::put_object(const l1::object_id& oid, builder_context& ctx) {
 ss::future<std::optional<partition_metadata>>
 reconciler::add_partition_to_object(
-  builder_context& ctx, const attached_partition& partition) {
+  builder_context& ctx, ss::shared_ptr<source> partition) {
```
I think should be renamed `source`, and the function should be `add_source_to_object`. I'm OK with saying "partition" in the logs, though. That'll be easier for people less familiar with the code to understand in the common (or only) case outside of tests.

rockwotj replied:
Yes thank you for catching. If you see others feel free to fix.

dotnwat reviewed (dismissed):
this all looks great.

i think the last commit can help with some kinds of shutdown slowness, but i suspect that it won't fix the hang we saw in the last RNOT run. my guess, posted from slack thread:

```
i stared at the logs for the shutdown hang for a while. the notable
thing is that the broker spends its last 30 seconds trying to elect
leaders for a ton of raft groups. it makes me wonder if:

* ducktape sends kill -15 to each broker sequentially
* 2 out of 3 shutdown
* the 3rd is blocked somewhere on a replicate with no timeout (which
* is precisely what reconciliation does via add_objects).

shutting down the frontend would probably solve that (its abort source
is used in replicate calls), but that shutdown happens after the reconciler.
```

we don't really need to block merging this on a guaranteed fix for the shutdown hang, just wanted to mention that analysis.

wdberkeley commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -596,20 +598,16 @@ ss::future<std::expected<void, reconcile_error>> reconciler::commit_objects(
+            kafka::offset lro = partition.metadata.last_offset;
+            auto it = corrected_next_offsets.find(tidp);
+            if (it != corrected_next_offsets.end()) {
+                // We want the previous offset, because that is what was last
```
Good catch.

rockwotj replied:
Unit tests caught it not me 😄

wdberkeley reviewed (dismissed):
LGTM. Terminology between "partition" and "source" is a little mixed up but I think that's OK. Ideally we'd use source generally in the code and partition only when we mean a source that is a partition, and in the logs I prefer "partition" because that's what it'll be outside of tests and it's immediately understandable to other developers and to users what that is, unlike source.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73006-->
<!--17b9eea3-ded1-4c34-9c10-450814ce8e2b-->
<!--27ec4175-0687-4fa6-93d1-a6dda7944e27-->
<!--adc26cb2-7e9c-4c3a-9534-7dcfa0c19999-->
<!--aa513966-58c4-4fc3-b63e-86f4a5464307-->
<!--title-->
## Retry command for [Build#73006](https://buildkite.com/redpanda/redpanda/builds/73006)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/follower_fetching_test.py::FollowerFetchingTest.test_with_leadership_transfers@{"fetch_from":"fetch-from-cloud-topic"}
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTxTest.test_write
```
<!--/body-->

dotnwat commented:
This failure occurred in the first and in the repeat run so perhaps it isn't just flaky?

https://ci-artifacts.dev.vectorized.cloud/redpanda/73006/01998366-dc1e-483e-a732-210e10dcac5e/vbuild/ducktape/results/2025-09-26--001/report.html

```
rptest.services.utils.BadLogLines: <BadLogLines nodes=docker-rp-13(2) 
example="ERROR 2025-09-26 01:00:11,983 [shard 0:main] reconciler - 
reconciler.cc:279 - Abandoning reconciliation loop because the L1 metastore operation failed">
```

rockwotj committed:
ct/stm_api: add logging for failures to replicate

rockwotj committed:
ct/reconciler: adjust logging

rockwotj committed:
ct/reconciler: remove the term partition from more places

Try to be more consistent with naming here and not introduce confusion
with our new source abstraction. Naming is hard however. Let's keep the
logs speaking partition for the rest of the team reading these logs.

rockwotj commented:
> This failure occurred in the first and in the repeat run so perhaps it isn't just flaky?
> 
> https://ci-artifacts.dev.vectorized.cloud/redpanda/73006/01998366-dc1e-483e-a732-210e10dcac5e/vbuild/ducktape/results/2025-09-26--001/report.html
> 
> ```
> rptest.services.utils.BadLogLines: <BadLogLines nodes=docker-rp-13(2) 
> example="ERROR 2025-09-26 01:00:11,983 [shard 0:main] reconciler - 
> reconciler.cc:279 - Abandoning reconciliation loop because the L1 metastore operation failed">
> ```

I added more logging to understand this. There is an error in `replicated_apply` in the stm_stm_api, but I'm not sure why

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73026-->
<!--e39838ed-0a6d-4007-9c8c-6880e470c1fa-->
<!--69d6f86f-abb3-4fee-b113-2dd2ee665562-->
<!--c91c4236-3c70-4ceb-a1e0-b458b65891cc-->
<!--title-->
## Retry command for [Build#73026](https://buildkite.com/redpanda/redpanda/builds/73026)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/follower_fetching_test.py::FollowerFetchingTest.test_with_leadership_transfers@{"fetch_from":"fetch-from-cloud-topic"}
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTxTest.test_write
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--4034b618-4c17-47d0-8515-d5fe05b5dd36-->
<!--TEST RESULTS-->
<!--31538796-faf9-4b33-9d18-8057da0b43d4-->
<!--TEST RESULTS-->
<!--eb15cb82-b6cd-4006-a35c-1e61d0ec29b1-->
<!--TEST RESULTS-->
<!--c39c828f-9f64-40f1-818f-1dc501777df2-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73026">build#73026</a></summary>


| test_class                           | test_method                                       | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                         |
|:-------------------------------------|:--------------------------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest            | test_write                                        | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/73026#0199840f-f573-428d-9032-f6cb194681dc | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                                   |
| EndToEndCloudTopicsTxTest            | test_write                                        | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/73026#01998410-5374-429e-b169-69343ed74982 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                                   |
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync                        | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/73026#01998410-537e-4d06-873e-b794ee849ff7 | FLAKY         | 20/21    | upstream reliability is '93.026941362916'. current run reliability is '95.23809523809523'. drift is -2.21115 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync        |
| ClusterRateQuotaTest                 | test_client_group_consume_rate_throttle_mechanism | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/73026#01998410-5375-471d-88ea-3fb89e089653 | FLAKY         | 13/21    | upstream reliability is '93.47234814143246'. current run reliability is '61.904761904761905'. drift is 31.56759 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_consume_rate_throttle_mechanism |
| ClusterRateQuotaTest                 | test_client_group_produce_rate_throttle_mechanism | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/73026#01998410-5377-44ac-b4d4-a67b626980ec | FLAKY         | 16/21    | upstream reliability is '81.72348484848484'. current run reliability is '76.19047619047619'. drift is 5.53301 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_produce_rate_throttle_mechanism |
| ClusterRateQuotaTest                 | test_client_response_throttle_mechanism           | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/73026#01998410-537a-4554-999d-199b0f49bc29 | FLAKY         | 12/21    | upstream reliability is '91.96956889264581'. current run reliability is '57.14285714285714'. drift is 34.82671 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism           |
| DataMigrationsApiTest                | test_mount_inexistent                             | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/73026#0199840f-f57f-4e19-a104-0bd8e5597b99 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS                | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_mount_inexistent                            |
| FollowerFetchingTest                 | test_with_leadership_transfers                    | {"fetch_from": "fetch-from-cloud-topic"}                                     | integration | https://buildkite.com/redpanda/redpanda/builds/73026#0199840f-f579-4d79-9242-a2dc47b8ef0a | FLAKY         | 9/21     | upstream reliability is '100.0'. current run reliability is '42.857142857142854'. drift is 57.14286 and the allowed drift is set to 50. The test should FAIL             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=FollowerFetchingTest&test_method=test_with_leadership_transfers                    |
| DisablingPartitionsTest              | test_disable                                      | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/73026#01998410-5375-471d-88ea-3fb89e089653 | FLAKY         | 12/21    | upstream reliability is '85.43461237274863'. current run reliability is '57.14285714285714'. drift is 28.29176 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DisablingPartitionsTest&test_method=test_disable                                   |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73071">build#73071</a></summary>


| test_class                           | test_method                | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                  |
|:-------------------------------------|:---------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest            | test_write                 | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/73071#019986e1-cdf2-43ee-94ef-0c0d644ab598 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                            |
| EndToEndCloudTopicsTxTest            | test_write                 | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/73071#019986e3-628f-440c-840c-70085b3a198e | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                            |
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/73071#019986e1-cdfa-4885-8f39-664444d92ad4 | FLAKY         | 20/21    | upstream reliability is '93.30289193302892'. current run reliability is '95.23809523809523'. drift is -1.9352 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| DatalakeCustomPartitioningTest       | test_many_partitions       | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                       | integration | https://buildkite.com/redpanda/redpanda/builds/73071#019986e1-cdf6-4b68-a9e2-4e6b28780aec | FLAKY         | 8/21     | upstream reliability is '100.0'. current run reliability is '38.095238095238095'. drift is 61.90476 and the allowed drift is set to 50. The test should FAIL           | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeCustomPartitioningTest&test_method=test_many_partitions             |
| PartitionMovementUpgradeTest         | test_basic_upgrade         | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/73071#019986e3-6296-4d55-9b49-e7e47b7512f2 | FLAKY         | 20/21    | upstream reliability is '99.77678571428571'. current run reliability is '95.23809523809523'. drift is 4.53869 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementUpgradeTest&test_method=test_basic_upgrade                 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73161">build#73161</a></summary>


| test_class                                   | test_method                                  | test_arguments                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    | test_history                                                                                                                                                           |
|:---------------------------------------------|:---------------------------------------------|:---------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest                    | test_write                                   | null                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73161#01999614-3f47-457d-98e4-e3d56556bc41 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                     |
| EndToEndCloudTopicsTxTest                    | test_write                                   | null                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73161#01999646-61ae-4591-82b6-07968fbdf7a0 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                     |
| RedpandaNodeOperationsSmokeTest              | test_node_ops_smoke_test                     | {"cloud_storage_type": 1, "mixed_versions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/73161#01999614-3f4e-4063-88a0-0e163652978e | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
| src/v/storage/tests/storage_e2e_fixture_test | src/v/storage/tests/storage_e2e_fixture_test |                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/73161#019995ee-25dd-4ce0-ad2d-5af07f279532 | FAIL          | 0/1      |                                                                                                                                                           |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73182">build#73182</a></summary>


| test_class                          | test_method    | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                     |
|:------------------------------------|:---------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| WriteCachingFailureInjectionE2ETest | test_crash_all | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/73182#019996eb-7350-45aa-947f-0c234e60fc9c | FLAKY         | 18/21    | upstream reliability is '89.20863309352518'. current run reliability is '85.71428571428571'. drift is 3.49435 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<!--/body-->

rockwotj commented:
> > This failure occurred in the first and in the repeat run so perhaps it isn't just flaky?
> > https://ci-artifacts.dev.vectorized.cloud/redpanda/73006/01998366-dc1e-483e-a732-210e10dcac5e/vbuild/ducktape/results/2025-09-26--001/report.html
> > ```
> > rptest.services.utils.BadLogLines: <BadLogLines nodes=docker-rp-13(2) 
> > example="ERROR 2025-09-26 01:00:11,983 [shard 0:main] reconciler - 
> > reconciler.cc:279 - Abandoning reconciliation loop because the L1 metastore operation failed">
> > ```
> 
> I added more logging to understand this. There is an error in `replicated_apply` in the stm_stm_api, but I'm not sure why

So it's because leadership moves and that replica is no longer leader when trying to replicate. I will change the logging for this error, but I wonder if we want to drop the error log here to warning because it's always possible the commit fails and it's fine to retry.

dotnwat commented:
> > > This failure occurred in the first and in the repeat run so perhaps it isn't just flaky?
> > > https://ci-artifacts.dev.vectorized.cloud/redpanda/73006/01998366-dc1e-483e-a732-210e10dcac5e/vbuild/ducktape/results/2025-09-26--001/report.html
> > > ```
> > > rptest.services.utils.BadLogLines: <BadLogLines nodes=docker-rp-13(2) 
> > > example="ERROR 2025-09-26 01:00:11,983 [shard 0:main] reconciler - 
> > > reconciler.cc:279 - Abandoning reconciliation loop because the L1 metastore operation failed">
> > > ```
> > 
> > 
> > I added more logging to understand this. There is an error in `replicated_apply` in the stm_stm_api, but I'm not sure why
> 
> So it's because leadership moves and that replica is no longer leader when trying to replicate. I will change the logging for this error, but I wonder if we want to drop the error log here to warning because it's always possible the commit fails and it's fine to retry.

Are you talking about L1 metastore leadership (e.g. committing) or the L0 CTP (e.g. advance the LRO)?

rockwotj committed:
ct/reconciler: add formatter for source::errc

rockwotj committed:
ct/l1/io: make errc formattable

rockwotj committed:
ct/reconciler: rework logging

Instead of every layer logging and having to correlate logs across
scheduling points, just accumulate error messages, then print logs only
in the top level `reconcile` method.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73071-->
<!--83488138-cf99-41bb-838e-2643e87ab4f7-->
<!--7b17fab1-ebc4-4ded-811e-35abd0bed103-->
<!--6940fb77-032e-41df-9393-ccdf1cb184e7-->
<!--title-->
## Retry command for [Build#73071](https://buildkite.com/redpanda/redpanda/builds/73071)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/custom_partitioning_test.py::DatalakeCustomPartitioningTest.test_many_partitions@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTxTest.test_write
```
<!--/body-->

dotnwat commented on src/v/cloud_topics/reconciler/reconciler.h:
```diff
@@ -35,7 +35,38 @@ class frontend;
+        return result;
+    }
+
+    // Mark this error as unexpected and not benign.
```
this comment is intended for `non_benign()`?

rockwotj replied:
yeah

dotnwat commented on src/v/cloud_topics/reconciler/reconciliation_consumer.cc:
```diff
@@ -35,8 +34,6 @@ reconciliation_consumer::operator()(model::record_batch batch) {
 
     // NOTE: Only data batches here. It's safe to use timestamps without
     //       checking the batch type.
-    _metadata.base_timestamp = std::min(
-      batch.header().first_timestamp, _metadata.base_timestamp);
```
why is base timestamp being removed (its removal is in the commit which renames partition -> source, so the commit reflects that not this removal which is probably misplaced?)

rockwotj replied:
Yeah I should have put that in another commit.

It's not used so it was removed, but I forgot to commit it seperately

dotnwat replied:
No worries

rockwotj commented:
Force push: fix conflicts


The test failing is a bit confusing to me because there are no changes to L0 here.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73161-->
<!--41fb82ff-d27a-495b-a8a7-47adf35faf02-->
<!--555ab930-5df8-4bf4-807d-5ef742012458-->
<!--title-->
## Retry command for [Build#73161](https://buildkite.com/redpanda/redpanda/builds/73161)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_topics/e2e_test.py::EndToEndCloudTopicsTxTest.test_write
```
<!--/body-->

dotnwat commented:
For the hanging Kgo consumer verifier, that looks like it's the TxE2E test. We should probably debug that separately--it seems unrelated to the changes in this PR.

rockwotj commented:
It's pretty consistently failing, so I think I'll need to debug here. I'm looking now

dotnwat commented:
Oh i see `set_last_reconciled_offset` so maybe that's affecting L0 reads

rockwotj committed:
rptest/ct: add trace logging to transaction test

We see some errors in this test, and this is useful to debug. Maybe we
can turn down logging later, but the volume in this test is tiny so I
think it's fine to leave.

rockwotj committed:
ct/reconciler: add logs for successful reconciliation

rockwotj committed:
ct/l0/ctp_stm: don't wastefully replicate a no-op LRO advance cmd

This actually fixes a real bug too, if reconcilation tries to reconcile
LRO `kafka::prev_offset(0)` which in `kafka::offset::min()` - this will
throw translation errors when the offset is applied to the STM. The fix
is to just not replicate offsets if it won't do anything, this prevents
this case. All other cases should be a bug.

rockwotj commented:
Okay found the issue, this was the log line that was coming up:

```
WARN  2025-09-29 17:50:34,316 [shard 0:main] raft - [group_id:4, {kafka/panda_topic/0}] state_machine_manager.cc:152 - [default][ctp_stm] error applying batch with base_offset: 345 - std::runtime_error (ntp {kafka/panda_topic/0}: data offset -9223372036854775808 is outside the translation range (starting at 0))
```

See the last commit for the fix. KgoVerifier was timing out because the STM was busted and trying to sync it on the fetch path always caused the fetch request to timeout
